lib/utils/normalizeOptions.js

```diff
@@ -4,7 +4,119 @@
   no-undefined
 */
 
+const path = require('path');
+const isAbsoluteUrl = require('is-absolute-url');
+const defaultTo = require('./defaultTo');
+
 function normalizeOptions(compiler, options) {
+  const firstWpOpt = compiler.compilers
+    ? compiler.compilers[0].options
+    : compiler.options;
+
+  // defaultTo only returns the second parameter if the first is null or undefined
+  firstWpOpt.mode = defaultTo(firstWpOpt.mode, 'development');
+
+  const devServerOptions = firstWpOpt.devServer || {};
+
+  // webpackOptions.devServer has the lowest priority here.
+  // This is a breaking change for CLI, but a desired one.
+
+  // For the API, it did not use webpackOptions.devServer in previous versions.
+  // With these changes, API users, optimally, should do 1 of 3 things:
+  // 1. Put all config in webpackOptions.devServer, then call: new Server(compiler, {});
+  // 2. Put all config in webpackOptions.devServer, then call: new Server(compiler, webpackOptions.devServer);
+  // 3. Put no config in webpackOptions.devServer, then call: new Server(compiler, options);
+
+  // These are the recommended usage practices,
+  // because it will cause no collision between webpackOptions.devServer and normal options
+
+  // not using Object.assign because we want the current options object to remain,
+  // but populate it with data from devServer if it does not already exist
+  Object.keys(devServerOptions).forEach((key) => {
+    if (options[key] === undefined) {
+      options[key] = devServerOptions[key];
+    }
+  });
+
+  if (!options.publicPath) {
+    // eslint-disable-next-line
```

Comment:
why?

Reply:
This was there originally, but I don't know why. eslint is fine without it.